api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -412,6 +413,29 @@ public String toString() {
     }
   }
 
+  public static class VariantType extends PrimitiveType {
```

Comment:
I don't think that variant should be either a primitive type or a nested type. There are at least a few places where making variant a primitive creates problems. For example, `Schema` checks that a field is a primitive to allow it as an identifier field and the `Identity` transform will bind to a type if it is a primitive. There are also places that check whether a type is nested and make decisions based on whether there are sub-fields.

Neither one of those seems appropriate to me so I'd suggest making this extend just `Type`. Then we can adjust the logic in those places without assuming either one and potentially introducing bugs when the assumption is bad.

Reply:
I have `isPrimitiveType`() to return false. I was debating on that - if VariantType extends PrimitiveType, some internal change would be easier. But concept-wise t should be from Type directly and let me change it.